third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -356,9 +370,42 @@ where
                         .charge_drop_frame(non_ref_vals.iter())
                         .map_err(|e| set_err_info!(current_frame, e))?;
 
-                    if RTTCheck::should_perform_checks() {
-                        self.check_return_tys(&mut current_frame)
+                    // If the returning function has runtime checks on, the return types
+                    // will be on the caller stack.
+                    let caller_has_rt_checks = self
+                        .call_stack
+                        .0
+                        .last()
+                        .map(|f| RTTCheck::should_perform_checks(&f.function.function))
+                        .unwrap_or_default();
+                    let callee_has_rt_checks =
+                        RTTCheck::should_perform_checks(&current_frame.function.function);
+                    if callee_has_rt_checks {
+                        self.check_return_tys::<RTTCheck>(&mut current_frame)
                             .map_err(|e| set_err_info!(current_frame, e))?;
+                        if !caller_has_rt_checks {
+                            // The callee has pushed return types, but they aren't used by
+                            // the caller, so need to be removed.
+                            self.operand_stack
```

Comment:
Hm, seems wasteful to push and then pop across this trust boundary. Can the callee not push anything if caller is not checking? Otherwise we can create more type substitutions and push to stack, and then drop